### PR TITLE
[shard] fix with_comms wrapper

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #69493

When added `with_comms` decorator with arguments, we added an `with_comms_decorator` inner function, `with_comms()` will refer to a function object, the added parentheses was necessary to use in test cases.

This PR fixes the `with_comms` wrapper behavior, to allow we both specify with/without arguments in test cases:
```
@with_comms
def test_case:
    ...
```
or
```
@with_comms(backend="gloo")
def test_case:
    ...
```

Differential Revision: [D32897555](https://our.internmc.facebook.com/intern/diff/D32897555/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### CODE REVIEW

File: torch/testing/_internal/distributed/_sharded_tensor/__init__.py
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/_sharded_tensor/__init__.py:75]: Wondering if we can just do what `dist_init` does: https://github.com/pytorch/pytorch/blob/master/torch/testing/_internal/dist_utils.py#L20 to keep wrappers like these consistent.

File: torch/testing/_internal/distributed/_sharded_tensor/__init__.py
Comment (pritamdamania87) [inline torch/testing/_internal/distributed/_sharded_tensor/__init__.py:75]: nit: Do we need the '*' here?

File: test/distributed/_sharded_tensor/test_sharded_tensor.py
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:1853]: Was this an unintended change?

File: test/distributed/_sharded_tensor/test_sharded_tensor.py
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:1894]: Don't we need to keep this validation for remote_shards?

File: test/distributed/_sharded_tensor/test_sharded_tensor.py
Comment (pritamdamania87) [inline test/distributed/_sharded_tensor/test_sharded_tensor.py:1910]: Is this test no longer relevant?

File: torch/distributed/_sharded_tensor/api.py
Comment (pritamdamania87) [inline torch/distributed/_sharded_tensor/api.py:443]: nit: Any reason to move this line?

**Review by pritamdamania87 — COMMENTED** [review]
Code changes look good, mostly had questions regarding some of the unit test changes.

File: test/distributed/_sharded_tensor/test_sharded_tensor.py
Comment (wanchaol) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:1910]: There were some issues when I try to run this test, mostly because if everywhere is none, we ended up having the `local_shards_device` be CPU, but we are calling `all_gather_object` on NCCL pg, let me see what I can do to make this legit.

File: test/distributed/_sharded_tensor/test_sharded_tensor.py
Comment (wanchaol) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:1894]: I think it's a bit tricky, the reason that I disabled this remote shards validation for this new_group test is that, we are essentially having two pg (default_pg, new_pg) for the test. When initializing sharded tensors, we only initialize (call the constructor) in ranks of the new_pg, but the `_init_rpc` call in `post_init()` will try to all_gather in the default_pg, in this case, rank 0 does not initialize(call constructor), so it does not have this all_gather, thus it hangs. 

https://github.com/pytorch/pytorch/blob/eb2a803406397e884c0d7999f782fa82e0204348/torch/distributed/_sharded_tensor/api.py#L220

I can think of a solution on this but not entirely sure the correct semantic:

we use `sharded_tensor.process_group.all_gather` instead of `rpc.api._all_gather` here, so that it only broadcast the sharded tensor ids in the new_pg, later we still share `all_tensor_ids` in default_pg world.

File: test/distributed/_sharded_tensor/test_sharded_tensor.py
Comment (wanchaol) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:1910]: just updated with removing the `local_shards_device` and check with `isinstance` instead, I think this should make it cleaner :)

File: test/distributed/_sharded_tensor/test_sharded_tensor.py
Comment (wanchaol) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:1894]: @pritamdamania87 I thought about this more and the solution i proposed might not work, as for those ranks who don't know about sharded tensor initialization, it does not have the `_register_remote_shards` or sharded tensor map on it at all. In this case, I am thinking that can we only broadcast and register remote shards in the `new_group` instead of the default_group?

https://github.com/pytorch/pytorch/blob/e963b436913f8324a12f8ff0987be7b4085020cb/torch/distributed/_sharded_tensor/api.py#L634

Comment (pritamdamania87) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:1894]: @wanchaol Yes I do feel we should do the all_gather only within the ranks of the `new_group`. Would this work out or are there other issues? I think we can potentially fix this in a follow PR to unblock landing this critical fix for the test suite.

**Review by pritamdamania87 — APPROVED** [review]

File: test/distributed/_sharded_tensor/test_sharded_tensor.py
Comment (wanchaol) [reply test/distributed/_sharded_tensor/test_sharded_tensor.py:1894]: yeah i think this would work out, i will work on a follow up PR after this land.

**Comment by wanchaol** [comment]
ci down because of sev https://github.com/pytorch/pytorch/issues/69722, not related to the PR

**Pull request closed without merging**
